### PR TITLE
deps!: support  cosmos-sdk-v45-ics-lsm integration

### PR DESCRIPTION
**Pull request by MSalopek**
## Description

Closes: #1183

Adapting ICS to support sdk-v45 with LSM support.

Mainly concerns about adapting the staking keeper interface and doing QA

Changes are state breaking.

# Instructions
When https://github.com/cosmos/cosmos-sdk/pull/16747 is merged and `cosmos-sdk v0.45.16-lsm-ics` release is cut perform the following actions:

1. update `go.mod`
Remove the `github.com/cosmos/cosmos-sdk => github.com/iqlusioninc/cosmos-sdk v0.45.17-0.20230803192832-aab739edf2ca` replace directive and use the correct `cosmos-sdk` version

2. run `go mod tidy`

3. commit the changes

4. cut `ICS v2.1.0`

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Provided a link to the relevant issue or specification
- [x] Reviewed "Files changed" and left comments if necessary 
- [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code

### CODE REVIEW

**Comment by MSalopek** [comment]
The issue below happens because of changes in cosmos-sdk logic regarding token shares.
```
simulate.go:299: error on block  26/100, operation (639/773) from x/authz:
        fee is not a subset of required fees; got 4304869cosmosvaloper15juvg98z0e67u5hu9jafp00dlsk08uqfqdnch9/1, required: 0stake: insufficient fee [/Users/msalopek/informal/gaia/x/globalfee/ante/fee.go:103]
        Comment: unable to deliver tx
--- FAIL: TestAppStateDeterminism (7.95s)
```

This is the function that introduces the change that breaks the `x/globalfee` module `AnteHandle`
```
// x/staking/types/tokenize_share_record.go [inqlusion's fork]
func (r TokenizeShareRecord) GetShareTokenDenom() string {
	return fmt.Sprintf("%s/%s", strings.ToLower(r.Validator), strconv.Itoa(int(r.Id)))
}
```

**Comment by MSalopek** [comment]
```
Starting the simulation from time Fri Jul 23 13:55:17 UTC 9948 (unixtime 251778981317)
Simulating... block 89/100, operation 0/350. simulation halted due to panic on block 89
Logs to writing to /Users/msalopek/.simapp/simulations/2023-07-13_18:29:09.log
--- FAIL: TestAppStateDeterminism (47.18s)
panic: division by zero [recovered]
	panic: division by zero [recovered]
	panic: division by zero

goroutine 37 [running]:
testing.tRunner.func1.2({0x101bf2ae0, 0x101e6c820})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1526 +0x1c8
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1529 +0x384
panic({0x101bf2ae0, 0x101e6c820})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/panic.go:884 +0x204
github.com/cosmos/cosmos-sdk/x/simulation.SimulateFromSeed.func2()
	/Users/msalopek/inqlusion/cosmos-sdk/x/simulation/simulate.go:146 +0xc8
panic({0x101bf2ae0, 0x101e6c820})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/panic.go:890 +0x258
math/big.nat.div({0x0?, 0x10314c5b8?, 0x20?}, {0x0?, 0x140131086e0?, 0x0?}, {0x140108f36c0?, 0x100a1f538?, 0x20?}, {0x0?, ...})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/math/big/natdiv.go:520 +0x260
math/big.(*Int).Quo(0x140131086e0, 0x1400261d058, 0x14013108640)
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/math/big/int.go:260 +0x64
github.com/cosmos/cosmos-sdk/types.Dec.Quo({0x101e72890?}, {0x0?})
	/Users/msalopek/inqlusion/cosmos-sdk/types/decimal.go:299 +0x74
github.com/cosmos/cosmos-sdk/x/staking/simulation.SimulateMsgTokenizeShares.func1(_, _, {{0x101e85f00, 0x1400011e018}, {0x101e90838, 0x1400f6632c0}, {{0x0, 0x0}, {0x1012ecb2e, 0x8}, ...}, ...}, ...)
	/Users/msalopek/inqlusion/cosmos-sdk/x/staking/simulation/operations.go:789 +0x734
github.com/cosmos/cosmos-sdk/x/simulation.createBlockSimulator.func1(_, _, {{0x101e85f00, 0x1400011e018}, {0x101e90838, 0x1400f6632c0}, {{0x0, 0x0}, {0x1012ecb2e, 0x8}, ...}, ...}, ...)
	/Users/msalopek/inqlusion/cosmos-sdk/x/simulation/simulate.go:290 +0x3ac
github.com/cosmos/cosmos-sdk/x/simulation.SimulateFromSeed({0x101e96440?, 0x14000ca0ea0}, {0x101e70360?, 0x14000126008}, 0x1?, 0x3d?, 0x101e658b0, {0x1400a4d4a00, 0x18, 0x20}, ...)
	/Users/msalopek/inqlusion/cosmos-sdk/x/simulation/simulate.go:182 +0x10e8
github.com/cosmos/gaia/v11/app_test.TestAppStateDeterminism(0x14000ca0ea0)
	/Users/msalopek/informal/gaia/app/sim_test.go:117 +0x750
testing.tRunner(0x14000ca0ea0, 0x101e66180)
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1576 +0x10c
created by testing.(*T).Run
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1629 +0x368
FAIL	github.com/cosmos/gaia/v11/app	47.622s
FAIL
make: *** [test-sim-nondeterminism] Error 1
```

**Comment by MSalopek** [comment]
**Errs in non-determinism tests***
```
--- FAIL: TestAppStateDeterminism (50.88s)
panic: division by zero [recovered]
	panic: division by zero [recovered]
	panic: division by zero

goroutine 36 [running]:
testing.tRunner.func1.2({0x106222b00, 0x10649c920})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1526 +0x1c8
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1529 +0x384
panic({0x106222b00, 0x10649c920})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/panic.go:884 +0x204
github.com/cosmos/cosmos-sdk/x/simulation.SimulateFromSeed.func2()
	/Users/msalopek/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/simulation/simulate.go:146 +0xc8
panic({0x106222b00, 0x10649c920})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/panic.go:890 +0x258
math/big.nat.div({0x0?, 0x107790a68?, 0x20?}, {0x0?, 0x1401acbc0a0?, 0x14000d32b60?}, {0x1401acb89c0?, 0x105041b08?, 0x20?}, {0x0?, ...})
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/math/big/natdiv.go:520 +0x260
math/big.(*Int).Quo(0x1401acbc0a0, 0x14003da1058, 0x1401acbc000)
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/math/big/int.go:260 +0x64
github.com/cosmos/cosmos-sdk/types.Dec.Quo({0x1064a2990?}, {0x0?})
	/Users/msalopek/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/types/decimal.go:299 +0x74
github.com/cosmos/cosmos-sdk/x/staking/simulation.SimulateMsgTokenizeShares.func1(_, _, {{0x1064b5fd0, 0x140001a0018}, {0x1064c0918, 0x1400f663580}, {{0x0, 0x0}, {0x105918cc9, 0x8}, ...}, ...}, ...)
	/Users/msalopek/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/staking/simulation/operations.go:789 +0x734
github.com/cosmos/cosmos-sdk/x/simulation.createBlockSimulator.func1(_, _, {{0x1064b5fd0, 0x140001a0018}, {0x1064c0918, 0x1400f663580}, {{0x0, 0x0}, {0x105918cc9, 0x8}, ...}, ...}, ...)
	/Users/msalopek/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/simulation/simulate.go:290 +0x3ac
github.com/cosmos/cosmos-sdk/x/simulation.SimulateFromSeed({0x1064c6500?, 0x14000d329c0}, {0x1064a0460?, 0x140001a8008}, 0x1?, 0x3d?, 0x1064959e0, {0x140085ed400, 0x23, 0x40}, ...)
	/Users/msalopek/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/simulation/simulate.go:182 +0x10e8
github.com/cosmos/gaia/v11/app_test.TestAppStateDeterminism(0x14000d329c0)
	/Users/msalopek/informal/gaia/app/sim_test.go:117 +0x750
testing.tRunner(0x14000d329c0, 0x1064962b0)
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1576 +0x10c
created by testing.(*T).Run
	/opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1629 +0x368
FAIL	github.com/cosmos/gaia/v11/app	51.446s
FAIL
make: *** [test-sim-nondeterminism] Error 1
```

```
simulate.go:299: error on block  26/100, operation (639/773) from x/authz:
        fee is not a subset of required fees; got 4304869cosmosvaloper15juvg98z0e67u5hu9jafp00dlsk08uqfqdnch9/1, required: 0stake: insufficient fee [/Users/msalopek/informal/gaia/x/globalfee/ante/fee.go:103]
        Comment: unable to deliver tx
--- FAIL: TestAppStateDeterminism (7.95s)
```

**Some issues with e2e tests (sovereign -> consumer)**
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xcc87d4]

goroutine 1 [running]:
math/big.(*Int).Sign(...)
	/usr/local/go/src/math/big/int.go:38
github.com/cosmos/cosmos-sdk/types.Dec.IsNegative(...)
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/types/decimal.go:211
github.com/cosmos/cosmos-sdk/x/staking/types.validateValidatorBondFactor({0x160bc20?, 0x0?})
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/staking/types/params.go:260 +0x44
github.com/cosmos/cosmos-sdk/x/params/types.Subspace.SetParamSet({{_, _}, _, {_, _}, {_, _}, {_, _, _}, ...}, ...)
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/params/types/subspace.go:245 +0x1a8
github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.SetParams(...)
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/staking/keeper/params.go:84
github.com/cosmos/cosmos-sdk/x/staking.InitGenesis({{0x1efc868, 0x4000050080}, {0x1f0a150, 0x400013d080}, {{0x0, 0x0}, {0x4000df4d19, 0x5}, 0x0, {0x3b832133, ...}, ...}, ...}, ...)
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/staking/genesis.go:35 +0x1ec
github.com/cosmos/cosmos-sdk/x/staking.AppModule.InitGenesis({{{0x1f10398, 0x4000fff5b0}}, {{0x1ee88f8, 0x4000c39e40}, {0x1f0a970, 0x4000fff5b0}, {0x1f00320, 0x40001525a0}, {0xffff5d1b9f40, 0x4000145e40}, ...}, ...}, ...)
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/x/staking/module.go:153 +0xdc
github.com/cosmos/cosmos-sdk/types/module.(*Manager).InitGenesis(_, {{0x1efc868, 0x4000050080}, {0x1f0a150, 0x400013d080}, {{0x0, 0x0}, {0x4000df4d19, 0x5}, 0x0, ...}, ...}, ...)
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/types/module/module.go:327 +0x1ec
github.com/cosmos/interchain-security/v2/app/sovereign.(*App).InitChainer(_, {{0x1efc868, 0x4000050080}, {0x1f0a150, 0x400013d080}, {{0x0, 0x0}, {0x4000df4d19, 0x5}, 0x0, ...}, ...}, ...)
	/interchain-security/app/sovereign/app.go:625 +0x1d0
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).InitChain(0x4000b516c0, {{0x3b832133, 0xedc1a47a9, 0x0}, {0x4000df4d19, 0x5}, 0x4000058680, {0x2d483d8, 0x0, 0x0}, ...})
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/baseapp/abci.go:63 +0x344
github.com/tendermint/tendermint/abci/client.(*localClient).InitChainSync(0x4000da1440, {{0x3b832133, 0xedc1a47a9, 0x0}, {0x4000df4d19, 0x5}, 0x4000058680, {0x2d483d8, 0x0, 0x0}, ...})
	/go/pkg/mod/github.com/cometbft/cometbft@v0.34.28/abci/client/local_client.go:272 +0x18c
github.com/tendermint/tendermint/proxy.(*appConnConsensus).InitChainSync(0x4000d08de0?, {{0x3b832133, 0xedc1a47a9, 0x0}, {0x4000df4d19, 0x5}, 0x4000058680, {0x2d483d8, 0x0, 0x0}, ...})
	/go/pkg/mod/github.com/cometbft/cometbft@v0.34.28/proxy/app_conn.go:77 +0x80
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(_, {{{0xb, 0x0}, {0x16307ee, 0x7}}, {0x4000df4d19, 0x5}, 0x1, 0x0, {{0x0, ...}, ...}, ...}, ...)
	/go/pkg/mod/github.com/cometbft/cometbft@v0.34.28/consensus/replay.go:319 +0xa9c
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0x4000ca3e30, {0x1f0d698, 0x4000570c30})
	/go/pkg/mod/github.com/cometbft/cometbft@v0.34.28/consensus/replay.go:268 +0x3a0
github.com/tendermint/tendermint/node.doHandshake({_, _}, {{{0xb, 0x0}, {0x16307ee, 0x7}}, {0x4000df4d19, 0x5}, 0x1, 0x0, ...}, ...)
	/go/pkg/mod/github.com/cometbft/cometbft@v0.34.28/node/node.go:329 +0x11c
github.com/tendermint/tendermint/node.NewNode(0x40001768c0, {0x1ef9160, 0x4000573900}, 0x4000c4dea0, {0x1ee3780, 0x4000b02570}, 0x1ee3240?, 0x40005781a0?, 0x4000c820b0, {0x1efd5f8, ...}, ...)
	/go/pkg/mod/github.com/cometbft/cometbft@v0.34.28/node/node.go:779 +0x3f0
github.com/cosmos/cosmos-sdk/server.startInProcess(_, {{0x0, 0x0, 0x0}, {0x1f19058, 0x4000d9aa50}, {0x0, 0x0}, {0x1f01ed8, 0x4000fff5b0}, ...}, ...)
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/server/start.go:280 +0x608
github.com/cosmos/cosmos-sdk/server.StartCmd.func2(0x4000ae7800?, {0x40005708f0?, 0x0?, 0xd?})
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/server/start.go:128 +0xf0
github.com/spf13/cobra.(*Command).execute(0x4000ae7800, {0x4000570820, 0xd, 0xd})
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x5e0
github.com/spf13/cobra.(*Command).ExecuteC(0x4000de6300)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x368
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:985
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0x400006c768?, {0x4000ff9120, 0x1c})
	/go/pkg/mod/github.com/iqlusioninc/cosmos-sdk@v0.45.17-0.20230711153639-a118c0fc9ad3/server/cmd/execute.go:36 +0x1a0
main.main()
	/interchain-security/cmd/interchain-security-sd/main.go:15 +0x38
```

**Comment by MSalopek** [comment]
Issue after backporting cancel-unbond operation handling and tests
```
➜  interchain-security git:(masa/test-ics-v2-with-cosmos-sdk-lsm) ✗ go run ./tests/e2e/... --happy-path-only
=============== running happy path only ===============
=============== building default testRun ===============
=============== started default tests ===============
running default: step 1 == StartChainAction
running default: step 2 == submitConsumerAdditionProposalAction
running default: step 3 == assignConsumerPubKeyAction
running default: step 4 == assignConsumerPubKeyAction
running default: step 5 == assignConsumerPubKeyAction
running default: step 6 == voteGovProposalAction
running default: step 7 == startConsumerChainAction
running default: step 8 == addIbcConnectionAction
running default: step 9 == addIbcChannelAction
running default: step 10 == delegateTokensAction
running default: step 11 == SendTokensAction
running default: step 12 == relayPacketsAction
running default: step 13 == SendTokensAction
running default: step 14 == assignConsumerPubKeyAction
running default: step 15 == relayPacketsAction
running default: step 16 == unbondTokensAction
running default: step 17 == relayPacketsAction
running default: step 18 == unbondTokensAction
running default: step 19 == relayPacketsAction
running default: step 20 == cancelUnbondTokensAction
2023/08/03 12:52:27 exit status 1
Error: unknown command "cancel-unbond" for "staking"
Usage:
  simd tx staking [flags]
  simd tx staking [command]

Available Commands:
  create-validator               create new validator initialized with a self-delegation to it
  delegate                       Delegate liquid tokens to a validator
  disable-tokenize-shares        Disable tokenization of shares
  edit-validator                 edit an existing validator account
  enable-tokenize-shares         Enable tokenization of shares
  redeem-tokens                  Redeem specified amount of share tokens to delegation
  redelegate                     Redelegate illiquid tokens from one validator to another
  tokenize-share                 Tokenize delegation to share tokens
  transfer-tokenize-share-record Transfer ownership of TokenizeShareRecord
  unbond                         Unbond shares from a validator
  unbond-validator               Unbond a validator
  validator-bond                 Mark a delegation as a validator self-bond

Flags:
  -h, --help   help for staking

Global Flags:
      --chain-id string     The network chain ID
      --home string         directory for config and data (default "/root/.interchain-security-p")
      --log_format string   The logging format (json|plain) (default "plain")
      --log_level string    The logging level (trace|debug|info|warn|error|fatal|panic) (default "info")
      --trace               print out full stack trace on errors

Use "simd tx staking [command] --help" for more information about a command.

exit status 1
```

File: legacy_ibc_testing/testing/app.go [hunk 1]
```diff
@@ -79,20 +79,23 @@ func SetupWithGenesisValSet(t *testing.T, appIniter AppIniter, valSet *tmtypes.V
...
+			DelegatorShares: sdk.OneDec(),
+			Description:     stakingtypes.Description{},
+			UnbondingHeight: int64(0),
+			UnbondingTime:   time.Unix(0, 0).UTC(),
+			Commission:      stakingtypes.NewCommission(sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec()),
+			// MinSelfDelegation: sdk.ZeroInt(),
 		}
 
 		validators = append(validators, validator)
+		// NOTE: @MSalopek
```
Comment (mpoke) [inline legacy_ibc_testing/testing/app.go:96]: That's because we don't have any ICS tests that are dealing with this flag.

**Review by mpoke — APPROVED** [review]
Nice work @MSalopek. LGTM

File: legacy_ibc_testing/testing/app.go [hunk 2]
```diff
@@ -79,20 +79,23 @@ func SetupWithGenesisValSet(t *testing.T, appIniter AppIniter, valSet *tmtypes.V
...
+			ConsensusPubkey: pkAny,
+			Jailed:          false,
+			Status:          stakingtypes.Bonded,
+			Tokens:          bondAmt,
+			DelegatorShares: sdk.OneDec(),
+			Description:     stakingtypes.Description{},
+			UnbondingHeight: int64(0),
+			UnbondingTime:   time.Unix(0, 0).UTC(),
+			Commission:      stakingtypes.NewCommission(sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec()),
+			// MinSelfDelegation: sdk.ZeroInt(),
```
Comment (p-offtermatt) [inline legacy_ibc_testing/testing/app.go:92]: Why is this commented out instead of removed outright?

**Review by p-offtermatt — APPROVED** [review]
LGTM, thanks for the work. Just one minor comment on a line that's commented out.

File: legacy_ibc_testing/testing/app.go (hunk 2 above)
Comment (mpoke) [reply legacy_ibc_testing/testing/app.go:92]: It was deprecated in SDK. Probably @MSalopek can give a better reasoning for commenting it out.

**Pull request merged**